processing/src/main/java/org/apache/druid/segment/join/HashJoinSegmentStorageAdapter.java

```diff
@@ -209,13 +209,36 @@ public Metadata getMetadata()
       @Nullable final QueryMetrics<?> queryMetrics
   )
   {
-    if (!Objects.equals(joinFilterPreAnalysis.getOriginalFilter(), filter)) {
-      throw new ISE(
-          "Filter provided to cursor [%s] does not match join pre-analysis filter [%s]",
+    JoinFilterPreAnalysis jfpa;
+    if (filter == null) {
+      jfpa = JoinFilterAnalyzer.computeJoinFilterPreAnalysis(
+          joinFilterPreAnalysisGroup,
+          JoinableClauses.fromList(clauses),
+          virtualColumns,
+          null,
+          joinFilterPreAnalysisGroup.isEnableFilterPushDown(),
+          joinFilterPreAnalysisGroup.isEnableFilterRewrite(),
+          joinFilterPreAnalysisGroup.isEnableRewriteValueColumnFilters(),
+          joinFilterPreAnalysisGroup.getFilterRewriteMaxSize()
+      );
+    } else {
+      jfpa = joinFilterPreAnalysisGroup.getAnalyses().computeIfAbsent(
           filter,
-          joinFilterPreAnalysis.getOriginalFilter()
+          (theFilter) -> {
+            return JoinFilterAnalyzer.computeJoinFilterPreAnalysis(
+                joinFilterPreAnalysisGroup,
+                JoinableClauses.fromList(clauses),
+                virtualColumns,
+                theFilter,
+                joinFilterPreAnalysisGroup.isEnableFilterPushDown(),
+                joinFilterPreAnalysisGroup.isEnableFilterRewrite(),
+                joinFilterPreAnalysisGroup.isEnableRewriteValueColumnFilters(),
+                joinFilterPreAnalysisGroup.getFilterRewriteMaxSize()
+            );
+          }
```

Comment:
I think a better abstraction is to hide all of this logic in the JoinFilterPreAnalysisGroup. We shouldn't expose `getAnalyses()` to the other classes.

Instead, consider exposing a function like `getPreAnalysisForFilter(Filter f)` - that provides the `jfpa` Users of the API don't need to worry about how it works as long as we guarantee that this function is thread safe.

Reply:
Also, since the pre-analysis computation is expensive, the way this is written, I think it's still possible for all the threads on the historical to attempt to compute the preAnalysis, so we could see a spike in CPU usage.

Reply:
> Also, since the pre-analysis computation is expensive, the way this is written, I think it's still possible for all the threads on the historical to attempt to compute the preAnalysis, so we could see a spike in CPU usage.

Sorry I misread the javadocs. ConcurrentHashMap blocks on an operation on the same key when a computation is on-going, so this shouldn't be an issue.

From the javadocs
> The entire method invocation is performed atomically, so the function is
> applied at most once per key.  Some attempted update operations
> on this map by other threads may be blocked while computation
> is in progress, so the computation should be short and simple,
> and must not attempt to update any other mappings of this map.

Reply:
I've added a `computeJoinFilterPreAnalysisIfAbsent` and `getAnalysis` method to `JoinFilterPreAnalysisGroup`